src/components/Users/CreateUserForm.tsx

```diff
@@ -92,9 +100,131 @@ interface Props {
+    </div>
+  );
+};
+
 export default function CreateUserForm({ onSubmitSuccess }: Props) {
   const { t } = useTranslation();
 
+  const userExistsEnums = {
+    idle: 0,
+    checking: 1,
+    exists: 2,
+    available: 3,
+  };
+
+  const [usernameExists, setUsernameExists] = useState<number>(
+    userExistsEnums.idle,
+  );
+  const [usernameInput, setUsernameInput] = useState("");
+
+  const { data, error, isLoading } = useQuery({
+    queryKey: ["checkUsername", usernameInput],
+    queryFn: query(userApi.checkUsername, {
+      pathParams: { username: usernameInput },
+      silent: true,
+    }),
+    enabled: validateUsername(usernameInput),
+  });
+
+  useEffect(() => {
+    if (!validateUsername(usernameInput)) {
+      setUsernameExists(userExistsEnums.idle);
+      return;
+    }
+
+    if (isLoading) {
+      setUsernameExists(userExistsEnums.checking);
+    } else if (error) {
+      if (error instanceof Error && "status" in error) {
+        const status = (error as any).status;
+
+        if (status === 409) {
+          setUsernameExists(userExistsEnums.exists);
+        } else if (status === 404) {
+          toast.error(
+            "Some error occurred while checking username availability. Please try again later.",
+          );
+          setUsernameExists(userExistsEnums.idle);
+        }
+      } else {
+        toast.error(
+          "Some error occurred while checking username availability. Please try again later.",
+        );
+        setUsernameExists(userExistsEnums.idle);
+      }
+    } else {
+      setUsernameExists(userExistsEnums.available);
+    }
+  }, [data, error, isLoading, usernameInput]);
+
+  const renderFeedback = () => {
+    switch (usernameExists) {
+      case userExistsEnums.checking:
+        return (
+          <div className="flex items-center gap-1">
+            <CareIcon icon="l-spinner" className="text-xl animate-spin" />
+            <span className="text-md text-gray-500">Checking username...</span>
+          </div>
+        );
+      case userExistsEnums.available:
+        return (
+          <div className="flex items-center gap-1">
+            <CareIcon
+              icon="l-check-circle"
+              className="text-xl text-primary-500"
+            />
+            <span className="text-md text-primary-500">
+              {t("username_available")}
+            </span>
+          </div>
+        );
+      case userExistsEnums.exists:
+        return (
+          <div className="flex items-center gap-1">
+            <CareIcon icon="l-times-circle" className="text-xl text-red-500" />
+            <span className="text-md text-red-500">
+              {t("username_not_available")}
+            </span>
+          </div>
+        );
+      default:
+        return null;
+    }
+  };
+
+  const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    const value = e.target.value;
+    setUsernameInput(value);
+    setUsernameExists(userExistsEnums.idle);
+  };
```

Comment:
to access zod's state, you can use `form.watch` instead of introducing a state variable.